cassiopeia/solver/CCPhyloSolver.py

```diff
@@ -0,0 +1,223 @@
+                internal_node_iter += 1
+            if not n.is_root():
+                tree.add_edge(n.up.name,n.name)
+
+        # find last split
+        midpoint = T.get_midpoint_outgroup()
+        root = T.get_tree_root()
+        if midpoint in root.children:
+            last_split = [root.name,midpoint.name]
+        else:
+            last_split = [root.name,root.children[0].name]
+        tree.remove_edge(last_split[0],last_split[1])
+
+        # root tree
+        tree = self.root_tree(tree,cassiopeia_tree.root_sample_name,last_split)
+
+        # remove root from character matrix before populating tree
+        if (
+            cassiopeia_tree.root_sample_name
+            in cassiopeia_tree.character_matrix.index
+        ):
+            cassiopeia_tree.character_matrix = (
+                cassiopeia_tree.character_matrix.drop(
+                    index=cassiopeia_tree.root_sample_name
+                )
+            )
+
+        # populate tree
+        cassiopeia_tree.populate_tree(tree,layer=layer)
+        cassiopeia_tree.collapse_unifurcations()
+
+        # collapse mutationless edges
+        if collapse_mutationless_edges:
+            cassiopeia_tree.collapse_mutationless_edges(
+                infer_ancestral_characters=True
+            )
+
+    def solve(
+        self,
+        cassiopeia_tree: CassiopeiaTree,
+        layer: Optional[str] = None,
+        collapse_mutationless_edges: bool = False,
+        logfile: str = "stdout.log",
+    ) -> None:
+        """Solves a tree for a general bottom-up distance-based solver routine.
+        If fast is set to True, this function will use the fast_solve method which
+        is wrapper around CCPhylo algorithms. Otherwise, it will default to the 
+        generic solve method in DistanceSolver. The function will update the `tree`
+        attribute of the input CassiopeiaTree.
+
+        Args:
+            cassiopeia_tree: CassiopeiaTree object to be populated
+            layer: Layer storing the character matrix for solving. If None, the
+                default character matrix is used in the CassiopeiaTree.
+            collapse_mutationless_edges: Indicates if the final reconstructed
+                tree should collapse mutationless edges based on internal states
+                inferred by Camin-Sokal parsimony. In scoring accuracy, this
+                removes artifacts caused by arbitrarily resolving polytomies.
+            logfile: File location to log output. Not currently used.
+        """
+
+        if self.fast:
+            self._fast_solve(cassiopeia_tree, layer, collapse_mutationless_edges, logfile)
+        else:
+            if self.__class__ == CCPhyloSolver:
+                raise NotImplementedError(
+                    "CCPhyloSolver does not implement solve. Please set fast to True" 
+                    "to use the fast_solve method. Or use a subclass of CCPhyloSolver."
```

Comment:
This is weird to me, especially since every method should implement `solve` without throwing an error. I think that this stems from the new structure of this PR (i.e., NeighborJoiningSolver inheriting from CCPhyloSolver). I wonder if we should consider getting rid of the CCPhyloSolver class and moving the `_fast_solve` logic to the `DistanceSolver` class. Then, if `fast=True` in the `DistanceSolver`, we'll divert the process to the `_fast_solver`; else we'll stay with the normal `solve` method.

Reply:
I like this solution. The CCPhyloSolver class is removed by c0c5ca926624067026e26d5483e49a18b3bc3f80